### ISSUE TITLE
Inner workings of nuclear

### ISSUE THREAD

**Issue by tindli** [optimizely/nuclear-js#13]
It seems that the react render mixin does NOT use shouldComponentUpdate().
This is pretty surprising because I really expected this... Instead, the components
call setState() if their state observable sees a difference in state...

So when I get this correctly, stateless components are true as the developer
does not have to deal with component state... everything is managed by nuclear; but
state is obviously there - spread all over the component tree.

Doesn't this hidden state all over the component tree trigger setState() more than really 
required or does nuclear intelligently detect the topmost component to call setState() for... 
leading to a single subtree re-rendering?

It would be great to have updated examples (Todos) to better experiment with nuclear...

Basically calling setState() deep in the component hierarchy is nice to perform
high performant updates on components - but as often propagated, state should
live as high in the component hierarchy as possible... (and therefore cause
a complete re-render of the whole app; something that immutable data makes
fast due to efficient use of shouldComponentUpdate()...) 

What is wrong with my understanding of nuclear???

**Comment by SimonDegraeve** [comment]
@jordangarcia any update on this, I am also curious about it.

**Comment by jordangarcia** [comment]
@tindli @SimonDegraeve These are very good questions, I will do my best to explain fully what is going on.

@tindli You're understanding of the use of Immutable Data and `componentShouldUpdate()` is not incorrect, in fact it is very spot on.  Immutable data gives us several benefits, one of the biggest in terms of performance is to check if any immutable data structure has changed with a `===` check (for map lookups this is Log32(n) runtime complexity)

Idiomatic usage of ReactJS would dictate that there is a single entry point for state in a component heirarchy and then that state is passed down the tree of components using props.  This is great as it allows sub components to be stateless and only a function of props.  Where it's not great is the need to "thread" data through your entire component heirarchy and you end up with components who should really be concerned about a certain part of the application state having to handle it in order to pass it to a child component.

With NuclearJS this paradigm is still available, you simply implement the `NuclearReactMixin` and `getDataBindings()` function on the root component and pass the data down as props to the child components.

However there is another strategy and that is to directly link component directly to whatever pieces of state they need and avoid the somewhat pointless passing data down the component heirarchy.  I prefer this way as it's easier to see the data dependencies of a component at the expense of making it agnostic to the underlying data system (using props is very generic and re-usable).  The point is NuclearJS gives you the option to do either.

**Performance**

When writing a ReactJS app with NuclearJS you should **almost** never be calling `setState` directly or a forcibly invoking a component render.  The benefits of Immutability live in the data layer.  What I mean by that is when a `dispatch` happens on the NuclearJS system it updates the state map by passing the action to all the `store` handlers and then get a new state map.  By binding the components directly to NuclearJS getters we can compare the new state map to the old state map and leverage the benefits of performant Immutable equality checks and then only call `setState()` on whatever components underlying state would have changed.  Thus we are already entering the React lifecycle with only the components that need re-rendering having a `setState` called on them directly, eliminating the need to do Virtual DOM diffing on a large part of the component hierarchy, since this was already determined in the NuclearJS dispatch lifecycle.

Hope this explanation helps!

**Comment by SimonDegraeve** [comment]
Indeed, it helps. Thank you.

**Comment by tindli** [comment]
@jordangarcia thanks for your detailed answer.

**Comment by mindjuice** [comment]
Great explanation.  This way of handling component state is actually what interested me most about Nuclear, and is why I've chosen it for my current project (just got started a few days ago).

The point about having the option to pass props is also key.  I plan to create several reusable components which may be used for both Nuclear and non-Nuclear projects in the future, so those will not use the NuclearReactMixin; they will just receive props.
